rfcs/react-components/convergence/motion-definition-n-apis.md

```diff
@@ -0,0 +1,1079 @@
+
+```tsx
+function MyButton() {
+  return <Button>Hello world!</Button>;
+}
+
+function App() {
+  return (
+    <>
+      {/* 🟢 works as Button handles classes properly */}
+      <Fade>
+        <Button />
+      </Fade>
+      {/* 🔴 fails as MyButton does not handle classes properly */}
+      <Fade>
+        <MyButton />
+      </Fade>
+    </>
+  );
+}
+```
+
+To make it work in this scenario, `MyButton` should handle the `className` prop:
+
+```tsx
+function MyButton() {
+  return <Button className={props.className}>Hello world!</Button>;
+}
+```
+
+The issue is that there is no way to enforce it in the API contract.
+
+</details>
+
+> Note: both CSS and Web Animations API options require to pass refs to elements, but that's easier to enforce as we check that a child uses `React.forwardRef()`.
+
+## Detailed design
+
+The proposed solution relies on the [Web Animations API](https://developer.mozilla.org/en-US/docs/Web/API/Web_Animations_API) and proposes to use the factories pattern to create React components for animating elements.
+
+<details>
+<summary>Why factories over hooks?</summary>
+
+To execute animations, we need a motion definition to create an animation. With hooks, this definition can be defined also inside a component, as illustrated below:
+
+```tsx
+// ⚠️ This is not proposed API, it's just an example
+
+function MyComponent() {
+  const motionRef = useAtomMotion({
+    keyframes: [{ opacity: 0 }, { opacity: 1 }],
+    duration: 2000,
+  });
+
+  return <div ref={motionRef}>Hello world!</div>;
+}
+```
+
+This approach presents a design issue: on every render, a new definition object is created and passed to the hook. At the same time, the motion must be a dependency to start the animation. Consequently, a new animation is created and started on every render, as depicted in the example below:
+
+```tsx
+function useAtomMotion(motion) {
+  const elementRef = React.useRef();
+
+  React.useEffect(() => {
+    const animation = elementRef.current.animate(motion.keyframes, {
+      duration: motion.duration,
+    });
+
+    return () => {
+      animation.cancel();
+    };
+  }, [motion]);
+}
+```
+
+With the factories pattern, the same problem is not possible, as a motion is defined outside a component, as shown below:
+
+```tsx
+function createAtomMotion(motion) {
+  return React.forwardRef((props, ref) => {
+    const elementRef = React.useRef();
+
+    React.useEffect(() => {
+      elementRef.current = elementRef.current.animate(motion.keyframes, {
+        duration: motion.duration,
+      });
+
+      return () => {
+        elementRef.current.cancel();
+      };
+      // `motion` does not need to be a dependency as it's defined outside the component
+    }, []);
+
+    return React.cloneElement(props.children, { ref: elementRef });
+  });
+}
+```
+
+</details>
+
+### Motion definition
```

Comment:
I think this RFC also needs to cover reduced motion use cases across the different proposed solutions.

For `useMotion` and RTG, the answer is fairly simple since CSS media queries could be applied to override the default animation styles.

For the WAPI motion definitions, I currently don't see a clear way to handle reduced motion from this proposal. I guess there should be some kind of matchMedia logic going on, but there could be multiple ways to integrate into the factory APIs, please show how that would be possible.

Reply:
There is a reduced motion implementation for perusal, pending an RFC writeup:

https://github.com/microsoft/fluentui/blob/master/packages/react-components/react-motions-preview/src/hooks/useIsReducedMotion.ts

https://github.com/microsoft/fluentui/blob/master/packages/react-components/react-motions-preview/src/factories/createAtom.ts#L47

Reply:
I would like to have a "developer toggle" for enabling reduced motion directly in JavaScript, for testing purposes. This would allow us to do things like show the same component with and without reduced motion side-by-side in Storybook. Whereas if reduced motion is only switchable globally, via the browser, we could not make this comparison simultaneously in the same page.

Reply:
It's also important to note that reduced motion does not necessarily mean that the duration of the animation is reduced. In an animation of an element where its size and color changes, the color change can still be desirable

Reply:
@ling1726 I added a section in 07139a0 that explains current state and possible solution.